pkgs/applications/science/biology/dcm2niix/default.nix

```diff
@@ -17,7 +24,23 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ cmake git ];
-  buildInputs = [ libyamlcpp ];
+  buildInputs =
+    lib.optional batchVersion [ libyamlcpp ]
+    ++ lib.optional withOpenJpeg [ openjpeg openjpeg.dev ]
+    ++ lib.optional withSystemZlib [ zlib ];
+
+  cmakeFlags =
+    lib.optional batchVersion [
+      "-DBATCH_VERSION=ON"
+      "-DYAML-CPP_DIR=${libyamlcpp}/lib/cmake/yaml-cpp"
+    ] ++ lib.optional withJpegLs [
+      "-DUSE_JPEGLS=ON"
+    ] ++ lib.optional withOpenJpeg [
+      "-DUSE_OPENJPEG=ON"
+      "-DOpenJPEG_DIR=${openjpeg}/lib/${openjpeg.pname}-${lib.versions.majorMinor openjpeg.version}"
+    ] ++ lib.optional withSystemZlib [
+      "-DZLIB_IMPLEMENTATION=System"
+    ];
```

Comment:
The "official" build process uses a fork of Cloudflare's zlib fork:

- https://github.com/rordenlab/dcm2niix/blob/002ebcdb9b2a87de7b883e9ddada3963a1cc2327/.travis.yml#L29
- https://github.com/rordenlab/dcm2niix/blob/2bf2e482aec8e9959c6bd8e833cdccba3607c617/SuperBuild/External-CLOUDFLARE-ZLIB.cmake

Rather than fall back to the included miniZ zlib implementation, I thought it'd be best to bridge the gap and use the zlib included with NixOS.

Reply:
Oh, if there is an actual use case for the vendored zlib we should name the option something like ``withCloudflareZlib`` or something that indicates the better performance.

Reply:
By default., `dcm2niix` is compiled with the vendored zlib. I don't want to change the default behavior. As an optional flag, CMake will fetch a fork of Cloudflare's zlib. However, there's no Nix package for that. System zlib is a decent middle ground, but I don't want to make it a default since it hasn't been tested upstream. What are your thoughts?

Reply:
Distros almost always devendor libaries because they are more often than not horrible out of date and though miss security patches which you also don't want to apply in several places.

> However, there's no Nix package for that.

We can add one for it. As far as I understood it, it should be a drop in replacement hence we also don't need to thorough testing.

Reply:
> Distros almost always devendor libraries because they are more often than not horrible out of date and though miss security patches which you also don't want to apply in several places.

This makes sense—100%. I suppose I fear that Miniz is a separate codebase than zlib. It's an alternative library that implements the zlib RFC. So, we're assuming liability by devendoring the library and switching to NixOS's zlib implementation. There could be subtle nuances between the Miniz zlib, NixOS's zlib, and this Cloudflare fork supported by the upstream repo. And, dcm2niix doesn't have a test suite, so there's no low-effort way to verify that things still work.

BTW, I appreciate the discussion and your time 🙂.

It sounds like our options are:

1) Assume the liability. Switch to NixOS's zlib. Hope that it's compatible with dcm2niix and users of the package don't encounter any nuance.
2) Bring in Cloudflare's Zlib to Nixpkgs and add an option called `withCloudflareZlib`. However, dcm2niix uses a frozen fork of Cloudflare's Zlib, so incompatibilities could crop up if we packaged a newer version in the distro.

I'm leaning towards #1 to get this out the door, but I'm also nervous about using this tool in our production environment with that enabled by default.

Reply:
I went with option one.